wallet/src/de/schildbach/wallet/ui/more/MasternodeKeysFragment.kt

```diff
@@ -0,0 +1,97 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.schildbach.wallet.ui.more
+
+import android.os.Bundle
+import android.view.View
+import androidx.core.os.bundleOf
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.viewModels
+import androidx.lifecycle.lifecycleScope
+import androidx.navigation.NavOptions
+import androidx.navigation.fragment.findNavController
+import androidx.recyclerview.widget.LinearLayoutManager
+import dagger.hilt.android.AndroidEntryPoint
+import de.schildbach.wallet.ui.CheckPinDialog
+import de.schildbach.wallet.util.Toast
+import de.schildbach.wallet_test.R
+import de.schildbach.wallet_test.databinding.FragmentMasternodeKeyTypesBinding
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.launch
+import org.dash.wallet.common.services.analytics.AnalyticsService
+import org.dash.wallet.common.ui.viewBinding
+import org.slf4j.LoggerFactory
+import javax.inject.Inject
+
+@FlowPreview
+@AndroidEntryPoint
+class MasternodeKeysFragment : Fragment(R.layout.fragment_masternode_key_types) {
+
+    companion object {
+        private val log = LoggerFactory.getLogger(MasternodeKeysFragment::class.java)
+    }
+    private val binding by viewBinding(FragmentMasternodeKeyTypesBinding::bind)
+
+    @Inject
+    lateinit var analytics: AnalyticsService
+    private val viewModel: MasternodeKeysViewModel by viewModels()
+    private val masternodeKeyTypeAdapter = MasternodeKeyTypeAdapter {
+        findNavController().navigate(
+            R.id.masternodeKeyChainFragment,
+            bundleOf("type" to it),
+            NavOptions.Builder()
+                .setEnterAnim(R.anim.slide_in_bottom)
+                .build(),
+        )
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.toolbar.setNavigationOnClickListener {
+            findNavController().popBackStack()
+        }
+        binding.title.setText(R.string.masternode_keys_title)
+        binding.keyTypeList.adapter = masternodeKeyTypeAdapter
+        binding.keyTypeList.layoutManager = LinearLayoutManager(activity)
+        binding.keyTypeList.setHasFixedSize(true)
+        if (viewModel.hasMasternodeKeys()) {
+            loadKeyTypes()
+        } else {
+            CheckPinDialog() { password ->
+                lifecycleScope.launch {
+                    viewModel.addKeyChains(password!!)
+                    loadKeyTypes()
+                }
+            }.showNow(childFragmentManager, "check-pin")
```

Comment:
this may be temporary, but we do need to add the authentication  (masternode) keychains to the Wallet somewhere.  We can and should also add them when a wallet is created.  But for existing wallets, we need to choose a place and time to add them: either here or the first time the app starts after being upgraded to the version that supports the Masternode Keys feature.

In the past, we asked for the user's pin before "upgrading" the wallet to support masternode keychains.  In DashPay, we don't do that when creating a username which must also "upgrade" the Wallet to support identity keychains.